### PR TITLE
TF-2384 [PART-1] Improvement notification iOS

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#2384 

## Task

- [x] Create NSE
- [x] Receive  payload data from FCM in NSE
- [x] Store `accountId, token, apiUrl, state.... to Keychain
- [x] Use data from Keychain to invoke api 
- [x] Show notification
- [x] Update app badger count 

## Environment

- Flutter version `3.16.0`

## Demo

https://github.com/linagora/tmail-flutter/assets/80730648/be244750-b1df-4e81-b152-b2c62214be5f

## Continuous

- PART II: https://github.com/linagora/tmail-flutter/pull/2408

### CODE REVIEW

File: ios/TwakeMailNSE/TwakeMailNSE.entitlements [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>com.apple.security.application-groups</key>
+	<array>
+		<string>group.com.linagora.teammail</string>
```
Comment (hoangdat) [inline ios/TwakeMailNSE/TwakeMailNSE.entitlements:7]: please make sure it is true

File: ios/TwakeMailNSE/NotificationService.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
+//
+//  NotificationService.swift
+//  TwakeMailNSE
+//
+//  Created by dab.dev on 12/12/2023.
+//
```
Comment (hoangdat) [inline ios/TwakeMailNSE/NotificationService.swift:6]: please remove it. We have git. No need to add it

File: ios/TwakeMailNSE/TwakeMailNSE.entitlements (hunk 1 above)
Comment (dab246) [reply ios/TwakeMailNSE/TwakeMailNSE.entitlements:7]: Of course it's right

**Comment by hoangdat** [comment]
- [ ] alert in APN payload

**Comment by hoangdat** [comment]
@dab246 please apply to branch of 3.16

File: contact/pubspec.lock [hunk 3]
```diff
@@ -297,10 +305,42 @@ packages:
...
+    version: "1.0.3"
+  flutter_image_compress_web:
+    dependency: transitive
+    description:
+      name: flutter_image_compress_web
+      sha256: da41cc3859f19d11c7d10be615f6a9dcf0907e7daffde7442bf4cc2486663660
       url: "https://pub.dev"
     source: hosted
-    version: "1.1.3"
+    version: "0.1.3+2"
```
Comment (hoangdat) [inline contact/pubspec.lock:343]: Why we need to downgrade it?

File: ios/Runner/AppDelegate.swift [hunk 4]
```diff
@@ -6,29 +6,19 @@ import flutter_local_notifications
 
 @UIApplicationMain
 @objc class AppDelegate: FlutterAppDelegate {
-    
-    /// Registers all pubspec-referenced Flutter plugins in the given registry.
-    static func registerPlugins(with registry: FlutterPluginRegistry) {
```
Comment (hoangdat) [inline ios/Runner/AppDelegate.swift:11]: Why we remove it?

Comment (dab246) [reply ios/Runner/AppDelegate.swift:11]: We previously created this function for reuse, but it is quite confusing and unnecessary. We will register directly for each Plugin when used

File: contact/pubspec.lock (hunk 3 above)
Comment (dab246) [reply contact/pubspec.lock:343]: This was an omission when we did issue #2346. We have upgraded the `flutter_image_compress` library in the `core` and main modules but have not updated it in other modules.

File: ios/Runner/AppDelegate.swift [hunk 5]
```diff
@@ -62,14 +52,13 @@ import flutter_local_notifications
...
-         // return false
-
-         // Proceed url handling for other Flutter libraries like uni_links
-         return super.application(app, open: url, options:options)
+        
+        return super.application(app, open: url, options:options)
+    }
+    
+    override func applicationDidBecomeActive(_ application: UIApplication) {
+        print("applicationDidBecomeActive")
```
Comment (hoangdat) [inline ios/Runner/AppDelegate.swift:60]: remove this log or only show in debug.

File: ios/Runner/AppDelegate.swift [hunk 6]
```diff
@@ -86,9 +75,49 @@ import flutter_local_notifications
         return URL(string: "ShareMedia-\(appDomain)://dataUrl=\(sharedKey)#text")
     }
     
+    // Receive displayed notifications for iOS 10 or later devices.
     override func userNotificationCenter(_ center: UNUserNotificationCenter,
                                          willPresent notification: UNNotification,
                                          withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
-        completionHandler(.alert) // shows banner even if app is in foreground
+        print("userNotificationCenter >> willPresent >> notification: \(notification)")
```
Comment (hoangdat) [inline ios/Runner/AppDelegate.swift:82]: idem

File: ios/Runner/AppDelegate.swift [hunk 7]
```diff
@@ -86,9 +75,49 @@ import flutter_local_notifications
...
+        if let notificationBadgeCount = notification.request.content.badge?.intValue, notificationBadgeCount > 0 {
+            let newBadgeCount = UIApplication.shared.applicationIconBadgeNumber + notificationBadgeCount
+            print("userNotificationCenter >> willPresent >> newBadgeCount: \(newBadgeCount)")
```
Comment (hoangdat) [inline ios/Runner/AppDelegate.swift:85]: idem

File: ios/Runner/AppDelegate.swift [hunk 8]
```diff
@@ -86,9 +75,49 @@ import flutter_local_notifications
...
+            updateAppBadger(currentBadgeCount: newBadgeCount)
+        }
+        
+        completionHandler(isAppForegroundActive() ? [.alert, .badge, .sound] : [])
```
Comment (hoangdat) [inline ios/Runner/AppDelegate.swift:89]: IMO, we should show notification if app is not in Foreground?

Comment (dab246) [reply ios/Runner/AppDelegate.swift:89]: Oh, That's right, I probably made a mistake here, I'll correct it.

**Review by hoangdat — APPROVED** [review]

**Pull request merged**